sql/sql_path.cc

```diff
@@ -0,0 +1,412 @@
+#include "sql_plugin.h"
+#include "sql_class.h"    // class THD
+#include "sp_head.h"      // class sp_name
+#include "lex_ident.h"
+#include "sql_db.h"
+#include "sql_path.h"
+
+static constexpr LEX_CSTRING cur_schema= {STRING_WITH_LEN(".")};
+
+Sql_path::Sql_path() :
+  m_schemas(m_schemas_array,
+            sizeof(m_schemas_array) / sizeof(m_schemas_array[0])),
+  m_count(0),
+  m_buffer(nullptr),
+  m_buffer_length(0)
+{
+}
+
+
+bool Sql_path::find_db_unqualified(THD *thd, const LEX_CSTRING &name,
+                                   const Sp_handler *sph,
+                                   Lex_ident_db_normalized *dbn_out,
+                                   sp_name **spname_out) const
+{
+  for (size_t i= 0; i < m_count; i++)
+  {
+    LEX_CSTRING schema= m_schemas[i];
+
+    /*
+      Resolve the schema name to the current database if required.
+    */
+    if (schema.length == cur_schema.length &&
+                        !memcmp(schema.str, cur_schema.str, cur_schema.length))
+      schema= thd->db;
+
+    if (!schema.str)
+      continue;
+
+    if (likely(!check_db_dir_existence(schema.str)))
+    {
+      const Lex_ident_db_normalized dbn=
+        thd->to_ident_db_normalized_with_error(schema);
+      if (dbn.str)
+      {
+        sp_name *spname= new (thd->mem_root) sp_name(dbn, name, false);
+        if (unlikely(!spname))
+          return true;
+
+        if (!sph->sp_find_routine_quick(thd, spname))
+        {
+          // found
+          if (dbn_out)
+            *dbn_out= dbn;
+          if (spname_out)
+            *spname_out= spname;
+
+          break;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+
+bool Sql_path::find_db_qualified(THD *thd,
+                                 sp_name *name,
+                                 const Sp_handler **sph,
+                                 Database_qualified_name *pkgname) const
+{
+  for (size_t i= 0; i < m_count; i++)
+  {
+    LEX_CSTRING schema= m_schemas[i];
+    /*
+      Resolve the schema name to the current database if required.
+    */
+    if (schema.length == cur_schema.length &&
+        !memcmp(schema.str, cur_schema.str, cur_schema.length))
+      schema= thd->db;
+
+    if (unlikely(!schema.str))
+      continue;
+    
+    if (likely(!check_db_dir_existence(schema.str)))
+    {
+      const Lex_ident_db_normalized dbn=
+        thd->to_ident_db_normalized_with_error(schema);
+      if (dbn.str)
+      {
+        if (!((*sph)->sp_find_qualified_routine(thd, dbn, name)))
+        {
+          pkgname->m_db= dbn;
+          pkgname->m_name= Lex_ident_routine(name->m_db);
+          *sph= (*sph)->package_routine_handler();
+          return name->make_package_routine_name(thd->mem_root, dbn,
+                                                 name->m_db, name->m_name);
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+
+Schema *Sql_path::find_first_internal_schema() const
```

Comment:
Let's move internal schema handling to a separate commit anda separate PR. Ok? Just to make commits a bit simpler and more focused.